index.js

```diff
@@ -873,6 +875,36 @@ class KeyringController extends EventEmitter {
       );
     }
   }
+
```

Comment:
There are two methods in which a new keyring is instantiated: `_restoreKeyring` and `addNewKeyring`.

I've extracted the logic so that we always instantiate, initialise (if available) and deserialize.

On the topic of deserialization, after looking at every single keyring I'm convinced that we don't have to pass the data at construction time as long as we call `deserialize` afterwards. `deserialize` is an asynchronous function, and we probably should stop calling it within the constructor.